### PR TITLE
Rewrite CLI to drop Boost dependency

### PR DESCRIPTION
**Pull request by jonpalmisc**
Ok, so here's my take on #61 to rewrite the CLI with a smaller argument parsing library. I chose to use https://github.com/adishavit/argh for this, which is a single header and I've added as a vendored dependency.

In the process I've also made a few changes—which I personally think are improvements—but might be controversial, and I'm willing to adjust if requested:

- All flags now have short flags:
  - `--cpu` can now also be passed as `-c`
  - `--features` can now also be passed as `-f`
  - `--address` can now also be passed as `-p`
  - `--labels` can now also be passed as `-l`
- Some flags have changed:
  - `--arch` has become `-t` and `--triple`
  - `-A` has been removed
  - `-D` has been renamed to `-d` (see below)
- The default "mode" is now to assemble. Use `-d` to request disassembly instead. I don't think it makes sense to have two flags to represent a binary option, where neither being passed is an invalid state.
- The formatting of assembled bytes when assembling has been made to look like an end-of-line comment, aligned past the end of each line (see example near end).

As a result, the help text has also been completely rewritten:

```
Usage: nyxstone [-t=<triple>] [-p=<pc>] [-d] <input>

Examples:
  # Assemble an instruction with the default architecture ('x86_64').
  nyxstone 'push eax'

  # Disassemble the bytes 'ffc300d1' as AArch64 code.
  nyxstone -t aarch64 -d ffc300d1

Options:
  -t, --triple=<triple>      LLVM target triple or alias, e.g. 'aarch64'
  -c, --cpu=<cpu>            LLVM CPU specifier, e.g. 'cortex-a53'
  -f, --features=<list>      LLVM architecture/CPU feature list, e.g. '+mte,-neon'
  -p, --address=<pc>         Initial address to assemble/disassemble relative to
  -l, --labels=<list>        Label-to-address mappings (used when assembling only)
  -d, --disassemble          Treat <input> as bytes to disassemble instead of assembly
  -h, --help                 Show this help and usage message

Notes:
  The '--target' parameter also supports aliases for common target triples:

     'x86_32' -> 'i686-linux-gnu'
     'x86_64' -> 'x86_64-linux-gnu'
     'armv6m' -> 'armv6m-none-eabi'
     'armv7m' -> 'armv7m-none-eabi'
     'armv8m' -> 'armv8m.main-none-eabi'
    'aarch64' -> 'aarch64-linux-gnueabihf'

  The CPUs for a target can be found with 'llc -mtriple=<triple> -mcpu=help'.
  The features for a target can be found with 'llc -mtriple=<triple> -mattr=help'.
```

Example of the changed output (notice opcode formatting):

```
$ nyxstone -t aarch64 'sub sp, sp, 0x70; stp x24, x23, [sp, 0x30]; ldr x8, [x8]'
	0x00000000: sub sp, sp, #112                 ; ff c3 10 d1 
	0x00000004: stp x24, x23, [sp, #48]          ; f8 5f 30 a9 
	0x00000008: ldr x8, [x8]                     ; 80 10 40 f9 
```

Let me know what you think of these changes. I understand this is quite a big change to drop on you guys—I'm willing to work with you to make this more mergeable as you all see fit. This probably needs some additional testing as well of course, but wanted to get a draft PR out before getting too far along.

But let's drop Boost once and for all. :)

### CODE REVIEW

File: examples/nyxstone-cli.cpp
Comment (stuxnot) [inline examples/nyxstone-cli.cpp:57]: `target` should be `triple`

File: examples/nyxstone-cli.cpp
Comment (stuxnot) [inline examples/nyxstone-cli.cpp:88]: I think it would be more in-line with existing code to have some fallible function `parse` (or similar) instead, which returns a `tl::expected<Options, std::string>`. Should also remove the need for the members `valid` and `error`.

File: examples/nyxstone-cli.cpp
Comment (stuxnot) [inline examples/nyxstone-cli.cpp:187]: nit: Failed _to_ decode bytes as hex

File: examples/nyxstone-cli.cpp
Comment (stuxnot) [inline examples/nyxstone-cli.cpp:20]: I think it is a bit unclear here (to a first time reader) why we return nothing if the `hex_string` does not hold an even number of characters. Having a comment along the lines of `This function parses instruction bytes, which we expect to always be typed out in full` or renaming the function to something like `decode_instruction_bytes` might make this clearer.

File: examples/nyxstone-cli.cpp
Comment (stuxnot) [inline examples/nyxstone-cli.cpp:33]: We previously allowed white spaces between hex bytes so that outputs from nyxstone could be freely copied back into it. I would like to keep this functionality.

**Review by stuxnot — CHANGES_REQUESTED** [review]
Thank you for your work, I really like the changes. The better flags and docs look really nice.

Apart from the comments I left, it would be great if you could update the README to reflect the changes. I also quite like the `vendor` subdirectory, since we vendored tl/expected as well, can you move the `expected.hpp` header file to the vendor subdirectory, too?

**Comment by jonpalmisc** [comment]
Glad you like the changes! I think I've resolved all of the things you mentioned with my latest commits.

**Comment by stuxnot** [comment]
Looks good!

Would you be willing to update the rust build.rs and the python setup.py to include the vendor directory? You will have to symlink `vendor` into the nyxstone and nyxstone-cpp subdirectories in the rust and python bindings folders. Additionally, to pass the CI, cppcheck and tidy will need to ignore the vendored code. If you need any help, let me know and maybe allow changes to your MR by maintainers, I can then implement this myself.

**Comment by jonpalmisc** [comment]
Hi, wanted to circle back to this. I haven't had as much time lately for open source stuff but am of course still interested in seeing this PR land.

I believe the PR allows edits by maintainers if you would still like to implement the changes mentioned in your last comment yourself. Otherwise, I will try to find some time to finish this up when I have time, but I'm not sure when that will be yet.

**Comment by stuxnot** [comment]
Alright, thank you for the update. I'm also a bit swamped at the moment, I'll try to finish this up soon.

**Comment by stuxnot** [comment]
Since I couldn't make changes to this PR, I opened a new one (#73) with the changes to fix the CI and the bindings. It should get merged as soon as there has been a review from the other maintainers. Thank you very much for your work! I really love the new and improved cli :smile:

**Pull request closed without merging**
